src/lib.rs

```diff
@@ -268,21 +271,95 @@ impl<'object, Endian> DebugInfo<'object, Endian>
+                            //  - if i == skiplist.len(), we scan from the last state machine
+                            //  - otherwise, the state machine from i-1 eventually encounters addr
+                            if i == 0 {
+                                unit.line_rows(&self.debug_line)
+                                    .chain_err(|| "cannot get line rows for unit")?
+                            } else {
+                                current = Some(skiplist[i - 1].2);
+                                rowi = (i + 1) * self.cache_every;
+                                skiplist[i - 1].1.clone()
+                            }
+                        }
+                    }
+                } else {
+                    // the skiplist was poisoned -- fall back to linear scan
+                    unit.line_rows(&self.debug_line)
+                        .chain_err(|| "cannot get line rows for unit")?
+                }
+            };
 
             // Now, find the last row before a row with a higher address than the one we seek.
-            let mut current = None;
-            while let Ok(Some((_, row))) = rows.next_row() {
-                if row.address() <= addr {
-                    if row.end_sequence() {
-                        current = None;
-                    } else {
-                        // Might be the right row, but we won't know until we see the next one.
-                        // The .clone is needed so we can keep iterating
-                        current = Some(*row);
+            let mut praddr = 0;
+            let mut skipseq = false;
+            while let Ok(Some((_, &row))) = rows.next_row() {
+                if row.end_sequence() {
+                    current = None;
+                    skipseq = false;
+                    continue;
+                }
+
+                if skipseq {
+                    continue;
+                }
+
+                let raddr = row.address();
+                if raddr == 0 {
+                    // NOTE: It is unclear if these rows are valid.
+                    continue;
+                }
+
+                if raddr < praddr {
+                    // NOTE:
+                    // We currently skip these sequences, but we *should* of course handle them
+                    // correctly. It's unclear how the interplay between this and the skiplist
+                    // should work.
+                    // unimplemented!();
+                    skipseq = true;
+                    continue;
+                }
+                praddr = raddr;
+
+                if raddr <= addr {
+                    // Might be the right row, but we won't know until we see the next one.
+                    // The .clone is needed so we can keep iterating
+                    current = Some(row);
+
+                    // We only cache non-end rows, because only there do we know that
+                    // .address() isn't 0.
```

Comment:
I don't understand this comment. I thought we handled end sequence and zero address already. Is it leftover from a prior version?